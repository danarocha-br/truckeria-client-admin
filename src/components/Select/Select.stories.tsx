import { Story, Meta } from '@storybook/react/types-6-0';
import { yupResolver } from '@hookform/resolvers/yup';
import { useForm, FormProvider } from 'react-hook-form';
import * as Yup from 'yup';
import { AnyObjectSchema } from 'yup';

import Select from '.';
import { FOOD_TYPE_OPTIONS } from 'constants/index';
import { useCallback } from 'react';

export default {
  title: 'Form Controls/Select',
  component: Select,
  parameters: {
    layout: 'centered',
  },
  args: {
    id: 'type',
    name: 'type',
    label: 'Select a food type',
    options: FOOD_TYPE_OPTIONS,
    loading: false,
    disabled: false,
    isMulti: false,
    isClearable: false,
  },
  argTypes: {
    id: {
      table: {
        category: 'Text',
      },
    },
    name: {
      table: {
        category: 'Text',
      },
    },
    label: {
      table: {
        category: 'Text',
      },
    },
    options: {
      table: {
        category: 'Text',
      },
    },
    isMulti: {
      table: {
        category: 'Modifiers',
      },
    },
    isClearable: {
      table: {
        category: 'Modifiers',
      },
    },
    loading: {
      table: {
        category: 'Modifiers',
      },
    },
    disabled: {
      table: {
        category: 'Modifiers',
      },
    },
  },
} as Meta;

const Template: Story = (args) => {
  type FormData = {
    description: string;
  };

  const initialValues: FormData = {
    description: '',
  };

  const SampleSchema = Yup.object().shape({
    description: Yup.string().min(5).required('Description is required'),
  });

  const methods = useForm<FormData>({
    defaultValues: { ...initialValues },
    mode: 'onTouched',
    resolver: yupResolver<AnyObjectSchema>(SampleSchema),
  });

  const {
    handleSubmit,
    // formState: { isSubmitting },
  } = methods;

  const onSubmit = useCallback(async (values: FormData) => {
    console.log(values);
  }, []);

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit(onSubmit)} className="w-96 space-y-4">
        <Select
          {...args}
          name="type"
          id="type"
          label="Select a food type"
          options={FOOD_TYPE_OPTIONS}
        />
      </form>
    </FormProvider>
  );
};

export const Single = Template.bind({});

export const isMulti = Template.bind({});
isMulti.args = {
  isMulti: true,
};
