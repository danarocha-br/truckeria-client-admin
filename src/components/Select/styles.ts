import styled, { css, DefaultTheme } from 'styled-components';
import tw from 'twin.macro';
import Select from 'react-select';
import { transparentize } from 'polished';

type ContainerProps = {
  isFocused: boolean;
  hasError: boolean;
  hasValue: boolean;
  isDisabled: boolean;
  readOnly?: boolean;
  isLoading: boolean | number;
};

type SelectProps = {
  isFocused: boolean;
  hasError: boolean;
  hasValue?: boolean;
  isDisabled: boolean;
  readOnly?: boolean;
  isLoading?: boolean | number;
};

const inputBorder = `
  border: 8px solid;
  border-top-width: 2.2em;
`;

const containerModifiers = {
  isFocused: (theme: DefaultTheme) => css`
    border-color: ${theme.form.focus};
  `,
  hasError: (theme: DefaultTheme) => css`
    border-color: ${theme.form.error};
  `,
  isDisabled: (theme: DefaultTheme) => css`
    ${tw`
      border
      border-dashed

    `}
    border-color: ${theme.text.subdued}60;
    background-color: ${theme.form.disabled};

    &:hover {
      border-color: ${theme.text.subdued}60;
    }
  `,

  readOnly: (theme: DefaultTheme) => css`
    label {
      ${tw`pl-7 pt-4`}
      background-color: ${theme.form.subdued};
    }

    &:hover {
      border-color: ${theme.background.default};
    }
  `,
};

const inputModifiers = {
  isFocused: (theme: DefaultTheme) => css`
    + label:before {
      ${inputBorder};
      border-top-width: 1.8em !important;
      border-color: ${theme.form.subdued} !important;
    }

    + label > span {
      transform: translate3d(-1em, -1.5em, 0) scale3d(0.8, 0.8, 1);
    }
  `,

  hasError: (theme: DefaultTheme) => css`
    border-color: ${theme.form.error};
  `,

  hasIcon: (theme: DefaultTheme) => css`
    border-color: ${theme.form.error};
  `,

  isDisabled: () => css`
    ${tw`cursor-not-allowed`};
    + label {
      ${tw`
        opacity-50
      `};
    }
  `,
};

export const Container = styled.span<ContainerProps>`
  ${({ theme, isFocused, hasError, isDisabled, readOnly, hasValue }) => css`
    ${tw`
      relative
      flex
      items-center
      w-full
      border
      transition
      duration-300
      h-20
      rounded-xs
    `}
    border-color: ${theme.background.default};
    min-width: 15rem;
    min-height: 5rem;

    &:hover {
      border-color: ${theme.form.focus};
    }

    ${isFocused && containerModifiers.isFocused(theme)};
    ${hasValue && containerModifiers.isFocused(theme)};
    ${hasError && containerModifiers.hasError(theme)};
    ${isDisabled && containerModifiers.isDisabled(theme)};
    ${readOnly && containerModifiers.readOnly(theme)};
  `}
`;

export const CustomSelect = styled(Select)<SelectProps>`
  ${({ theme, isFocused, hasError, hasValue, isDisabled }) => css`
    ${tw`
        w-full
        h-auto
        border-none
        rounded-xs
    `};
    position: absolute !important;

    .c-select {
      &__control {
        ${tw`
          flex
          border-none
          text-base
          z-10
          appearance-none
          focus:outline-none
          disabled:cursor-not-allowed
          disabled:opacity-40
          w-full
          h-full
          rounded-xs
          shadow-none
        `};

        color: ${theme.text.default};
        padding: 1.24em 0.5em 0;
        background: transparent !important;
      }

      &--is-focused {
        ${tw`
          border-none
          outline-none
          shadow-none
        `};
      }

      &--menu-is-open {
        z-index: 1999 !important;
      }

      &__menu {
        ${tw`
          rounded-xs
          mt-1
        `};
        z-index: 1999 !important;
        background-color: ${theme.surface.base.subdued};
      }

      &__menu-list {
        ${tw`
            p-0
            m-0
          `};
      }

      &__option {
        &:hover {
          ${tw`
            bg-blue-300
          `};
        }

        &--is-focused {
          ${tw`
            bg-blue-300
          `};
        }

        &--is-selected {
          ${tw`
            bg-blue-500
          `};
        }

        &:first-child {
          ${tw`
            rounded-t-xs
          `};
        }
        &:last-child {
          ${tw`
            rounded-b-xs
          `};
        }
      }

      &__indicator-separator {
        ${tw`hidden`};
      }

      &__single-value {
        color: ${theme.text.default};
      }

      &__input-container {
        color: ${theme.text.default};
      }

      &__multi-value {
        ${tw`
          border
          border-brand
          transition-colors
          mr-2
        `};
        background: transparent;

        &:hover {
          background-color: ${transparentize(0.8, theme.colors.brand)};
        }

        &__label {
          ${tw`
          mx-1
        `};
          color: ${theme.text.default};
        }
      }
      &__dropdown-icon {
        ${tw`transition-colors`};
        color: ${theme.text.default};

        &:hover {
          color: ${theme.text.subdued};
        }
      }
      &__clear-icon {
        ${tw`transition-colors`};
        color: ${theme.text.default};
        border-radius: 2px;

        &:hover {
          ${tw`text-red-500 bg-red-100`};
        }
      }
    }
    ${isFocused && inputModifiers.isFocused(theme)};
    ${hasError && inputModifiers.hasError(theme)};
    ${hasValue && inputModifiers.isFocused(theme)};
    ${isDisabled && inputModifiers.isDisabled()};
  `}
`;
