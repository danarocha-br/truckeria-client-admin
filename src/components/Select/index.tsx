import { InputHTMLAttributes, useCallback, useState } from 'react';
import { useFormContext, useController } from 'react-hook-form';
import {
  components,
  DropdownIndicatorProps,
  ClearIndicatorProps,
} from 'react-select';
import { AiOutlineLoading3Quarters, AiFillCaretDown } from 'react-icons/ai';
import { HiExclamationCircle, HiX } from 'react-icons/hi';

import * as S from './styles';
import { Error, Label } from 'components/TextInput/styles';
import tokens from '../../../tailwind.config';

type OptionType = {
  value: string | number;
  label: string;
};

export type SelectProps = {
  label?: string;
  id: string;
  name: string;
  disabled?: boolean;
  loading?: boolean;
  readOnly?: boolean;
  isMulti?: boolean;
  isClearable?: boolean;
  options: OptionType[];
} & InputHTMLAttributes<HTMLSelectElement>;

const Select = ({
  name,
  id,
  label,
  options,
  disabled = false,
  loading = false,
  isMulti = false,
  isClearable = false,
  defaultValue,
  ...rest
}: SelectProps) => {
  const { control, getValues } = useFormContext();

  const {
    field: { ref, onChange, ...inputProps },
    formState: { dirtyFields, errors },
  } = useController({
    name,
    control,
    defaultValue,
  });

  /**
   * Get UI States
   */
  const [isFocused, setFocus] = useState(!!dirtyFields[name]);

  const handleInputFocus = useCallback(() => {
    setFocus(true);
  }, []);

  const handleInputBlur = useCallback(() => {
    // if (!watchedInput || watchedInput.length > 0) {
    setFocus(false);
  }, []);

  /** Custom components */
  const DropdownIndicator = (props: DropdownIndicatorProps) => (
    <components.DropdownIndicator {...props}>
      <AiFillCaretDown size={14} className="c-select__dropdown-icon" />
    </components.DropdownIndicator>
  );

  const ClearIndicator = (props: ClearIndicatorProps) => {
    const {
      innerProps: { ref, ...restInnerProps },
      children = <HiX className="c-select__clear-icon" />,
    } = props;
    return (
      <div {...restInnerProps} ref={ref} style={{ padding: '0px 5px' }}>
        {children}
      </div>
    );
  };

  // const NoOptionsMessage = (props) => (
  //   <components.NoOptionsMessage {...props} />
  // );

  return (
    <>
      <div className="flex flex-col w-full rounded-sm">
        <S.Container
          isFocused={isFocused}
          hasError={errors[name] ? true : false}
          isDisabled={disabled || loading}
          isLoading={Number(loading)}
          hasValue={isMulti ? getValues(name)?.length : !!dirtyFields[name]}
        >
          <S.CustomSelect
            {...inputProps}
            {...rest}
            inputRef={ref}
            name={name}
            id={id}
            placeholder=""
            instanceId={id}
            classNamePrefix="c-select"
            options={options}
            isFocused={isFocused}
            hasError={errors[name] ? true : false}
            hasValue={!!dirtyFields[name] || !!defaultValue}
            isDisabled={disabled || loading}
            // value={value}
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            onChange={(e: KeyboardEvent | EventTarget | any) => {
              onChange(e);
            }}
            onFocus={handleInputFocus}
            onBlur={handleInputBlur}
            isClearable={isClearable}
            isMulti={isMulti}
            components={{
              DropdownIndicator,
              ClearIndicator,
            }}
          />
          <Label
            aria-labelledby={label}
            htmlFor={id}
            hasError={errors[name] ? true : false}
            style={{
              zIndex: -1,
            }}
          >
            <span>{label}</span>

            {errors[name] ? (
              <HiExclamationCircle
                name="alert"
                size={20}
                color={tokens.theme.colors.red[500]}
                className="c-input__error-icon"
                aria-hidden="true"
              />
            ) : null}

            {loading && (
              <AiOutlineLoading3Quarters
                color="white"
                className="animate-spin absolute top-1 right-2"
                aria-hidden="true"
              />
            )}
          </Label>
        </S.Container>

        {errors[name] ? (
          <Error role="alert">{errors[name].message}</Error>
        ) : null}
      </div>
    </>
  );
};

export default Select;
