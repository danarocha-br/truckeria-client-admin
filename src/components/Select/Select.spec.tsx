import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useForm, FormProvider } from 'react-hook-form';
import { renderHook } from '@testing-library/react-hooks';

import { renderWithTheme } from 'utils/tests/helpers';

import Select from '.';

describe('<Select />', () => {
  const { result } = renderHook(() => useForm());
  const methods = result.current;

  it('should render the Select', () => {
    const { container } = renderWithTheme(
      <FormProvider {...methods}>
        <form>
          <Select
            label="Label"
            id="field"
            name="field"
            options={[{ label: '1', value: '1' }]}
          />
        </form>
      </FormProvider>
    );

    expect(screen.getByRole('combobox')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Is accessible by tab', () => {
    renderWithTheme(
      <FormProvider {...methods}>
        <form>
          <Select
            label="Label"
            id="field"
            name="field"
            options={[{ label: '1', value: '1' }]}
          />
        </form>
      </FormProvider>
    );

    const input = screen.getByRole('combobox');
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).toHaveFocus();
  });
});
