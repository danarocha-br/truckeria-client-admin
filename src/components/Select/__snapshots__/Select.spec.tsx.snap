// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render the Select 1`] = `
.c0 {
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,-webkit-transform,filter,backdrop-filter;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  -webkit-transition-duration: 300ms;
  transition-duration: 300ms;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  border-width: 1px;
  height: 5rem;
  border-radius: 0.4rem;
  border-color: #FFFFFF;
  min-width: 15rem;
  min-height: 5rem;
}

.c0:hover {
  border-color: #FF8D4D;
}

.c1 {
  width: 100%;
  height: auto;
  border-style: none;
  border-radius: 0.4rem;
  position: absolute !important;
}

.c1 .c-select__control {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-style: none;
  font-size: 1.1rem;
  z-index: 10;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  height: 100%;
  border-radius: 0.4rem;
  --tw-shadow: 0 0 #0000;
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  color: #20242A;
  padding: 1.24em 0.5em 0;
  background: transparent !important;
}

.c1 .c-select__control:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.c1 .c-select__control:disabled {
  cursor: not-allowed;
  opacity: 0.4;
}

.c1 .c-select--is-focused {
  border-style: none;
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-shadow: 0 0 #0000;
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
}

.c1 .c-select--menu-is-open {
  z-index: 1999 !important;
}

.c1 .c-select__menu {
  border-radius: 0.4rem;
  margin-top: 0.25rem;
  z-index: 1999 !important;
  background-color: #B8C4C8;
}

.c1 .c-select__menu-list {
  padding: 0px;
  margin: 0px;
}

.c1 .c-select__option:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(41,178,176,var(--tw-bg-opacity));
}

.c1 .c-select__option--is-focused {
  --tw-bg-opacity: 1;
  background-color: rgba(41,178,176,var(--tw-bg-opacity));
}

.c1 .c-select__option--is-selected {
  --tw-bg-opacity: 1;
  background-color: rgba(33,142,141,var(--tw-bg-opacity));
}

.c1 .c-select__option:first-child {
  border-top-left-radius: 0.4rem;
  border-top-right-radius: 0.4rem;
}

.c1 .c-select__option:last-child {
  border-bottom-left-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
}

.c1 .c-select__indicator-separator {
  display: none;
}

.c1 .c-select__single-value {
  color: #20242A;
}

.c1 .c-select__input-container {
  color: #20242A;
}

.c1 .c-select__multi-value {
  border-width: 1px;
  --tw-border-opacity: 1;
  border-color: rgba(255,141,77,var(--tw-border-opacity));
  -webkit-transition-property: background-color,border-color,color,fill,stroke;
  transition-property: background-color,border-color,color,fill,stroke;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  margin-right: 0.5rem;
  background: transparent;
}

.c1 .c-select__multi-value:hover {
  background-color: rgba(255,141,77,0.2);
}

.c1 .c-select__multi-value__label {
  margin-left: 0.25rem;
  margin-right: 0.25rem;
  color: #20242A;
}

.c1 .c-select__dropdown-icon {
  -webkit-transition-property: background-color,border-color,color,fill,stroke;
  transition-property: background-color,border-color,color,fill,stroke;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  color: #20242A;
}

.c1 .c-select__dropdown-icon:hover {
  color: #636F73;
}

.c1 .c-select__clear-icon {
  -webkit-transition-property: background-color,border-color,color,fill,stroke;
  transition-property: background-color,border-color,color,fill,stroke;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  color: #20242A;
  border-radius: 2px;
}

.c1 .c-select__clear-icon:hover {
  --tw-text-opacity: 1;
  color: rgba(224,82,82,var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(247,213,213,var(--tw-bg-opacity));
}

.c2 {
  font-weight: 500;
  text-align: left;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
  padding-top: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  border-radius: 0.4rem;
  color: #20242A;
  background-color: #ECEFF4;
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2:before {
  border-style: solid;
  position: absolute;
  top: 0px;
  left: 0px;
  right: 0px;
  width: 100%;
  height: 100%;
  content: '';
  border-color: #ECEFF4;
  -webkit-transition: border-width 0.3s,border-color 0.5s;
  transition: border-width 0.3s,border-color 0.5s;
}

.c2 span {
  -webkit-transform: var(--tw-transform);
  -ms-transform: var(--tw-transform);
  transform: var(--tw-transform);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  position: relative;
  top: -0.25rem;
  margin-left: 0.75rem;
  -webkit-transition-duration: 300ms;
  transition-duration: 300ms;
  -webkit-transform-origin: 0% 50%;
  -ms-transform-origin: 0% 50%;
  transform-origin: 0% 50%;
  text-rendering: geometricPrecision;
}

.c2 svg.c-input__icon {
  -webkit-transform: var(--tw-transform);
  -ms-transform: var(--tw-transform);
  transform: var(--tw-transform);
  margin-left: 0.25rem;
  -webkit-transition-duration: 300ms;
  transition-duration: 300ms;
  margin-top: -8px;
}

.c2 svg.c-input__error-icon {
  position: absolute;
  right: 0.5rem;
  top: 0.25rem;
}

.c2 input:disabled + label:before {
  border-color: #E05252;
}

<form>
  <div
    class="flex flex-col w-full rounded-sm"
  >
    <span
      class="c0"
    >
      <div
        class="c1 css-b62m3t-container"
        id="field"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-field-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class="c-select__control css-1s2u09g-control"
        >
          <div
            class="c-select__value-container css-319lph-ValueContainer"
          >
            <div
              class="c-select__placeholder css-14el2xx-placeholder"
              id="react-select-field-placeholder"
            />
            <div
              class="c-select__input-container css-14dclt2-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-controls="react-select-field-listbox"
                aria-describedby="react-select-field-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                aria-owns="react-select-field-listbox"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class="c-select__input"
                id="react-select-field-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="c-select__indicators css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class="c-select__indicator-separator css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="c-select__indicator c-select__dropdown-indicator css-tlfecz-indicatorContainer"
            >
              <svg
                class="c-select__dropdown-icon"
                fill="currentColor"
                height="14"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="field"
          type="hidden"
          value=""
        />
      </div>
      <label
        aria-labelledby="Label"
        class="c2"
        for="field"
        style="z-index: -1;"
      >
        <span>
          Label
        </span>
      </label>
    </span>
  </div>
</form>
`;
